### ISSUE TITLE
A bit different update implementation from the Algorithm 1 from the paper.

### ISSUE THREAD

**Issue by Junyoungpark** [watchernyu/REDQ#6]
Hi, I found the implementation of code and the pseudocode in the paper is somewhat different.  

The algorithm box in the paper said; Update the critics `G` times and update the actor `once`. However, the implementation updates critics and actors `G` times using the same sampled batches. Can you clarify which is the one intended? I believe the code implementation is the real one you've done for producing the reported results.

**Comment by watchernyu** [comment]
@Junyoungpark Thank you for posting the issue! Maybe I can clarify this:

For the newest version of the code, if you go to
`REDQ/redq/algos/redq_sac.py`
and look at line 213, you will find: 

```
            """policy and alpha loss"""
            if ((i_update + 1) % self.policy_update_delay == 0) or i_update == num_update - 1:
                # get policy loss
```

Note the policy loss is used only inside this If statement. (which is different from the Q loss)

So basically the `policy_update_delay` here (default value is 20) will control how many policy updates we do per data point. Note in this case, the if statement evaluates to True only when i_update+1 == 20. So basically you take 20 Q updates, and then only 1 policy update, which is the same as reported in the paper. In the paper, the results in the main paper are produced with this default value. 

In the appendix of our paper, we also have results on what happens when we remove this "policy delay" (basically setting `policy_update_delay` to be 1), and we get similar results, but with slightly higher variance. Since the results are similar, and take less policy update means less computation, so by default we take only 1 policy update per datapoint. 

Hope that clarifies the issue! Let us know if you have further questions! :D

**Comment by Junyoungpark** [comment]
Thanks for the clarification 👍 . 

May I ask one thing more about REDQ?

Reusing `obs_tensor` already used to train Qs -- the last step of the Q training -- to train the actor and alpha is also intended? I think it is an intended design choice such that the policy can be trained where Qs are already trained, and consequently, the actor training step can exploit more accurate value estimates during training.

**Comment by watchernyu** [comment]
That is a good question! Actually we are not sure if it's "intended"... because this particular implementation of REDQ is build on top of the SAC algorithm, and we tried to keep most of the algorithmic components the same as the SAC algorithm (including most of the hyperparameters) so that our comparisons are fair. This choice of using the same batch to update Q and policy has been seen in many previous papers, including SAC, so we kept it the same way, and I'm not quite sure who started it first or why. 

My guess is probably this is not really sth that is specially designed, but maybe people write it this way just for the simplicity. And I'm not sure if the best method is to use the same batch or different batches. Using the same batch might have the benefit you mentioned... however, this might also bring more correlation to the Q and the actor, which might lead to a bit more bias accumulation? 

Looks like an interesting research question, if you are interested maybe you can try investigate it further :)

**Comment by sweetice** [comment]
Thanks for the clear code. I really enjoy reading the paper and the code.
I have a question. What would happen when increasing the policy update frequency? I change the policy update frequency from updating (20 times critic optimization then 1 time policy) to (2 times critic optimization then 1 time policy optimization). I conduct experiments on pybullet suite. I find that the performance is not consistent with the original paper.  
As we all know, pybullet is generally considered hard than mujoco suite.

**Comment by watchernyu** [comment]
@sweetice Thanks for posting the question! If you are running on different environments, it is possible that the best hyperparameters will be different. You can check out the ablation study on page 23 of our paper's appendix (https://arxiv.org/pdf/2101.05982.pdf) where we show that having more policy updates in general leads to a bit more variance, and that affect performance a bit (though different in each environment). 

A similar study can be found in the TD3 paper, where they show that policy delay can be helpful and stabilize training. 

So if you are running on harder environments, and also using more policy updates, it is possible that it might lead to more training instability. Hope that answers the question!

**Comment by sweetice** [comment]
> @sweetice Thanks for posting the question! If you are running on different environments, it is possible that the best hyperparameters will be different. You can check out the ablation study on page 23 of our paper's appendix (https://arxiv.org/pdf/2101.05982.pdf) where we show that having more policy updates in general leads to a bit more variance, and that affect performance a bit (though different in each environment).
> 
> A similar study can be found in the TD3 paper, where they show that policy delay can be helpful and stabilize training.
> 
> So if you are running on harder environments, and also using more policy updates, it is possible that it might lead to more training instability. Hope that answers the question!

Thanks for your kind reply!
